test/test_autodiff_lab.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+# Owner(s): ["module: unknown"]
+
+import torch
+from torch.autograd import gradcheck, gradgradcheck
+
+class AttentionLab(torch.autograd.Function):
+    """Define the forward and backward pass"""
+    @staticmethod
+    def forward(ctx, q, k, v):
+        x = torch.matmul(q, k.transpose(0, 1))
+        a = torch.tanh(x)
+        o = torch.matmul(a, v)
+        ctx.save_for_backward(q, k, v, a)
+        return o, a
+
+
+    @staticmethod
+    def backward(ctx, grad_o, grad_a):
+        q, k, v, a = ctx.saved_tensors
+
+        grad_q = grad_k = grad_v = None
```

Comment:
Also FYI (no change necessary), setting them all to `None` at the start is useful if you intend to check `needs_input_grad[idx]` later. See: https://pytorch.org/docs/stable/notes/extending.html#extending-autograd

This is similar to the case in c++ where you accept `grad_input_mask` argument in your backward function and decide what to compute based on that.

Both of these mechanisms are opt-in and are for improving performance.